state/protocol/badger/validity.go

```diff
@@ -111,3 +117,161 @@ func validCommit(commit *flow.EpochCommit, setup *flow.EpochSetup) error {
+	if identity.Role != expectedRole {
+		return engine.NewInvalidInputErrorf("expected node %x to have identity %s but got %s", nodeID, expectedRole, identity.Role)
+	}
+
+	// check if the identity has a stake
+	if identity.Stake == 0 {
+		return engine.NewInvalidInputErrorf("node has zero stake (%x)", identity.NodeID)
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) identityForNode(blockID flow.Identifier, nodeID flow.Identifier) (*flow.Identity, error) {
+	// get the identity of the origin node
+	identity, err := v.state.AtBlockID(blockID).Identity(nodeID)
+	if err != nil {
+		if protocol.IsIdentityNotFound(err) {
+			return nil, engine.NewInvalidInputErrorf("unknown node identity: %w", err)
+		}
+		// unexpected exception
+		return nil, fmt.Errorf("failed to retrieve node identity: %w", err)
+	}
+
+	return identity, nil
+}
+
+func (v *receiptValidator) verifySignature(receipt *flow.ExecutionReceipt, nodeIdentity *flow.Identity) error {
+	id := receipt.ID()
+	valid, err := v.verifier.Verify(id[:], receipt.ExecutorSignature, nodeIdentity.StakingPubKey)
+	if err != nil {
+		return fmt.Errorf("failed to verify signature: %w", err)
+	}
+
+	if !valid {
+		return engine.NewInvalidInputErrorf("Invalid signature for (%x)", nodeIdentity.NodeID)
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) verifyChunksFormat(result *flow.ExecutionResult) error {
+	for index, chunk := range result.Chunks.Items() {
+		if uint(index) != chunk.CollectionIndex {
+			return fmt.Errorf("invalid CollectionIndex, expected %d got %d", index, chunk.CollectionIndex)
+		}
+
+		if chunk.BlockID != result.BlockID {
+			return fmt.Errorf("invalid blockID, expected %s got %s", result.BlockID, chunk.BlockID)
+		}
+	}
+
+	// we create one chunk per collection, plus the
+	// system chunk. so we can check if the chunk number matches with the
+	// number of guarantees plus one; this will ensure the execution receipt
+	// cannot lie about having less chunks and having the remaining ones
+	// approved
+	requiredChunks := 1 // system chunk: must exist for block's ExecutionResult, even if block payload itself is empty
+
+	index, err := v.index.ByBlockID(result.BlockID)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return err
+		}
+		// reaching this line means the block is empty, i.e. it has no payload => we expect only the system chunk
+	} else {
+		requiredChunks += len(index.CollectionIDs)
+	}
+
+	if result.Chunks.Len() != requiredChunks {
+		return fmt.Errorf("invalid number of chunks, expected %d got %d", requiredChunks, result.Chunks.Len())
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) verifyExecutionResult(result *flow.ExecutionResult) error {
+	prevResult, err := v.results.ByID(result.PreviousResultID)
+	if err != nil {
+		return fmt.Errorf("no previous result ID")
+	}
+
+	block, err := v.state.AtBlockID(result.BlockID).Head()
+
+	if err != nil {
+		return fmt.Errorf("no block found %s %w", result.BlockID, err)
+	}
```

Comment:
to reduce dependency between the different methods (rather, in which order they have to be executed), I would suggest we also handle the possibility of `storage.ErrNotFound` error.
Thereby, there is no implicit assumption about in which order we call the verification methods `verifySignature`, `verifyChunksFormat`, `verifyExecutionResult`